pkg/core/statesync/module.go

```diff
@@ -459,6 +537,76 @@ func (s *Module) AddMPTNodes(nodes [][]byte) error {
 	return nil
 }
 
+// AddContractStorageData adds a batch of key-value pairs for storage-based sync.
+func (s *Module) AddContractStorageData(kvs []storage.KeyValue, syncHeight uint32, expectedRoot util.Uint256) error {
+	if s.mode == MPTBased {
+		panic("contract storage items are not expected in MPT-based mode")
+	}
+	oldStage := s.syncStage
+	s.lock.Lock()
+	defer func() {
+		if s.syncStage != oldStage {
+			s.notifyStageChanged()
+		}
+	}()
+	defer s.lock.Unlock()
+
+	if s.syncStage&headersSynced == 0 || s.syncStage&mptSynced != 0 || expectedRoot.Equals(s.localTrie.StateRoot()) {
+		return errors.New("contract storage items were not requested")
+	}
+	if syncHeight != s.syncPoint {
+		return fmt.Errorf("invalid sync height: expected %d, got %d", s.syncPoint, syncHeight)
+	}
+	if len(kvs) == 0 {
+		return fmt.Errorf("key-value pairs are empty")
+	}
+
+	var (
+		prefix = TemporaryPrefix(s.dao.Version.StoragePrefix)
+		batch  = make(map[string][]byte, len(kvs))
+	)
+	for _, kv := range kvs {
+		s.dao.Store.Put(append([]byte{byte(prefix)}, kv.Key...), kv.Value)
+		batch[string(kv.Key)] = kv.Value
+	}
+	mptBatch := mpt.MapToMPTBatch(batch)
```

Comment:
Use this:
```
	var (
		prefix = TemporaryPrefix(s.dao.Version.StoragePrefix)
		batch  = make(map[string][]byte, len(kvs))
	)
	for _, kv := range kvs {
		batch[string(append([]byte{byte(prefix)}, kv.Key...))] = kv.Value
	}
	s.dao.Store.PutChangeSet(nil, batch)
	mptBatch := mpt.MapToMPTBatch(batch)
```
`s.dao.Store.Put` takes mutex over storage every time you call it, we don't need it.